commons/zenoh-buffers/src/zslicemut.rs

```diff
@@ -0,0 +1,89 @@
+//
+// Copyright (c) 2023 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+use core::ops::{Deref, DerefMut, Range};
+
+/*************************************/
+/*             ZSLICEMUT             */
+/*************************************/
+pub struct ZSliceMut<'a> {
+    slice: &'a mut [u8],
+    start: &'a mut usize,
+    end: &'a mut usize,
+}
+
+impl<'a> ZSliceMut<'a> {
+    pub(crate) fn new(slice: &'a mut [u8], start: &'a mut usize, end: &'a mut usize) -> Self {
+        Self { slice, start, end }
+    }
+
+    pub fn range(&self) -> Range<usize> {
+        *self.start..*self.end
+    }
+
+    pub fn set_range(&mut self, range: Range<usize>) -> bool {
+        match self.slice.len() >= range.end {
```

Comment:
`start` should also be checked. With the current check new `start` could be smaller than the original `start.

Reply:
start is just an offset from the underlying buffer ptr. It is always safe to set it to 0, as we just rewind to the very beginning of the buffer.
But my error is in end index check, yes, I will fix it

Reply:
Rechecked this. This code is correct, as it validates that the upper bound of the range fits the underlying buffer. And with lower bound: it is safe to rewind up to 0 position